src/os/bluestore/BlueStore.cc

```diff
@@ -8287,56 +8287,189 @@ int BlueStore::repair_omap_upgrade()
-        string s2 = key.substr(p2 - 8, 8);
-        if (s1 != s2) {
-          continue;
-        }
-        txn->rmkey(PREFIX_PERPG_OMAP, key);
-        if (key[p2] == '.') {
-          auto new_key = key;
-          new_key.erase(p1, p2 - p1);
-          dout(0) << __func__ << " " << pretty_binary_string(key)
-                  << " -> " << pretty_binary_string(new_key) << dendl;
-          txn->set(PREFIX_PERPG_OMAP, new_key, val);
-        } else {
-          dout(0) << __func__ << " " << pretty_binary_string(key)
-                  << " rm " << dendl;
-        }
-        cnt++;
-        if (cnt >= 1024) {
-          key = it->key();
-          db->submit_transaction_sync(txn);
-          it->lower_bound(key);
-          txn = db->get_transaction();
-          cnt = 0;
-        }
+	dout(20) << __func__ << " key=" << pretty_binary_string(key) << dendl;
+	it->next();
+	ceph_assert(key.length() >= 8 + 4 + 8 + 1);
+	const char* p = key.c_str();
+	uint64_t pool;
+	unsigned hash;
+	uint64_t nid;
+	p = _key_decode_u64(p, &pool);
+	p = _key_decode_u32(p, &hash);
+	p = _key_decode_u64(p, &nid);
+	if (pool == prev_pool && hash == prev_hash && nid == prev_nid) {
+	  if (expect_next_object) {
+	    dout(10) << __func__ << " next object expected, but old continued" << dendl;
+	    problems++;
+	  }
+	} else {
+	  if (!expect_next_object) {
+	    // we should continue same object
+	    dout(10) << __func__ << " next objext unexpected" << dendl;
+	    problems++;
+	    expect_header_copy = false;
+	  }
+	  prev_pool = pool;
+	  prev_hash = hash;
+	  prev_nid = nid;
+	}
+	char c = *p;
+	p++;
+
+	if (expect_header_copy) {
+	  // copy of header should look like:
+	  // pppppppphhhhnnnnnnnn.ppppppppnnnnnnnn-   (for PERPOOL->PERPG failed upgrade)
+	  // pppppppphhhhnnnnnnnn.nnnnnnnn-           (for BULK->PERPG failed upgrade)
+
+	  // we expect copy only right after valid header
+	  expect_header_copy = false;
+	  if (key.length() == 8 + 4 + 8 + 1 + 8 + 8 + 1) {
+	    uint64_t pool_copy;
+	    uint64_t nid_copy;
+	    char c1;
+	    const char* p1 = p;
+	    p1 = _key_decode_u64(p1, &pool_copy);
+	    p1 = _key_decode_u64(p1, &nid_copy);
+	    c1 = *p1;
+	    if (c == '.' && pool_copy == pool && nid_copy == nid && c1 == '-') {
+	      // bingo - delete
+	      dout(15) << __func__ << " rmkey " << pretty_binary_string(key) << dendl;
+	      txn->rmkey(PREFIX_PERPG_OMAP, key);
+	      cnt += 1000;
+	      continue;
+	    }
+	  }
+	  if (key.length() == 8 + 4 + 8 + 1 + 8 + 1) {
+	    uint64_t nid_copy;
+	    char c1;
+	    const char* p1 = _key_decode_u64(p, &nid_copy);
+	    c1 = *p1;
+	    if (c == '.' && nid_copy == nid && c1 == '-') {
+	      // bingo - delete
+	      dout(15) << __func__ << " rmkey " << pretty_binary_string(key) << dendl;
+	      txn->rmkey(PREFIX_PERPG_OMAP, key);
+	      cnt += 1000;
+	      continue;
+	    }
+	  }
+	  dout(10) << __func__ << " expected copy of header. not found" << dendl;
+	  problems++;
+	}
+
+	if (c == '~') {
```

Comment:
Just FYI: there is no guarantee that tail is always present. The following PR introduced it in O+ releases and there were backports to N and M: https://github.com/ceph/ceph/pull/27627
But OMAPs produced by Luminous and early N/M releases might still lack it.
IIUC your state machine critically depends on that record presence hence letting you know as this might be a showstopper. 
But I'm still uncertain whether we need a feature-proof implementation for that recovery